deepspeed/runtime/zero/stage3.py

```diff
@@ -912,6 +892,43 @@ def __init__(self,
         if dist.get_rank(group=self.dp_process_group) == 0:
             see_memory_usage(f"After initializing ZeRO optimizer", force=False)
 
+    @staticmethod
+    def defragment(tensors: List[Tensor]) -> Tensor:
```

Comment:
Awesome!

Reply:
its mostly just to defragment tensor data thats created during initialization and kept throughout the training process (parameter partitions). We found that this helped with memory pressure a bit. 

defragmentation of more transient memory allocations was already handled before i started making changes, and i didn't see much opportunity to further reduce fragmentation without creating a lot of overhead

Reply:
Could we make this accessible to users down the road at training time? I can surely see situations where a user will want to run this occasionally at their own schedule. So that the default run doesn't do it, but it can be configured to defrag periodically or say via a low mem counter trigger?

This is an awesome feature!

Reply:
@jfc4050, thanks for the explanation. I concur with @stas00 that this should ideally be promoted to a utility outside ZeRO. So can you please annotate with a TODO so that we don't forget.

Reply:
works for me, added TODO